### ISSUE TITLE
Add ways to customize outer loop flow

### ISSUE THREAD

**Issue by apaleyes** [amzn/emukit#25]
It should be possible to customize the loop flow without rewriting it. For instance, if a user wants to store some benchmarking information between iterations, they should be able to add it into the existing loop without redefining it.

Right now we have `self.custom_step()` at the end of each iteration, which does the job to some extent, but obviously isn't very flexible and does not cover for many scenarios.

**Comment by javiergonzalezh** [comment]
We have 4 main steps in the grand loop: update model, run policy (collect points), evaluate objective and update the state. Maybe an idea is to have a custom call for each one of these components instead of having just one for the entire loop.

**Comment by marpulli** [comment]
If the only need for this at the minute is to save benchmarking information, should we have an optional `ModelScorer` class as part of the outer loop that does this instead? I'm imagining:

```
class ModelScorer:
    def score_model(self, model)
        stuff here...
```

I'm not convinced about the name yet, I'll keep thinking about that

**Comment by apaleyes** [comment]
No, the idea was to give more flexibility in general. At the moment however benchmarking is the only thing that comes to mind.

Implementation-wise i was thinking of something like that:

       self.loop_starting()
       while not stopping_condition.should_stop(self.loop_state):
            self.loop_iteration_started()

            self.model_updating()
            self.model_updater.update(self.loop_state)
            self.model_updated()

            self.next_points_computing()
            new_x = self.candidate_point_calculator.compute_next_points(self.loop_state)
            self.next_points_computed()

            self.objective_executing()
            results = user_function.evaluate(new_x)
            self.objective_executed()

            self.loop_state_updating()
            self.loop_state.update(results)
            self.loop_state_updated()

            self.loop_iteration_finished()

        self.loop_finished()

And when users have a loop object, they would subscribe to these events, maybe with something as simple as `my_loop.loop_state_updated = my_event_handler()`. And these events may have arguments too, say loop state, maybe more.

Hopefully the idea is clear, although of course this is very verbose - we don't really need that many events. Let me know what you think.

**Comment by apaleyes** [comment]
First step in this direction is implemented by @marpulli  in #53 . We will add more customization to the loop as we find use cases for it.
